doc: update zlib doc (#31665)
nodejs/node: doc/api/zlib.md

> Stability: 2 - Stable

The `zlib` module provides compression functionality implemented using Gzip,
Deflate/Inflate, and Brotli.

It can be accessed using:

```js
const zlib = require('zlib');
```

Compression and decompression is built around the Node.js [Streams API][].

Compressing or decompressing a stream (such as a file) can be accomplished by
piping the source stream through a `zlib` `Transform` stream into a destination
stream:

```js
const { createGzip } = require('zlib');
const { pipeline } = require('stream');
const {
  createReadStream,
  createWriteStream
} = require('fs');

const gzip = createGzip();
const source = createReadStream('input.txt');
const destination = createWriteStream('input.txt.gz');

pipeline(source, gzip, destination, (err) => {
  if (err) {
    console.error('An error occurred:', err);

>     process.exit(1);

Reviewer: Why forcing the process to exit in examples?

Author: To emphasize that the error needs to be handled somehow, and typically crashing is safer than continuing unless things are handled properly.

Reviewer: Wouldn't this example exit anyway if an error occurs? If so wouldn't it be better to use `process.exitCode = 1` instead of forcing exit if the example wants to set the exit code to 1?

Author: That assumption may not hold if the sample code is cut and pasted into some other code.

Reviewer: Hmm it doesn't make much sense. What makes you think that the error should be handled by forcing the process to exit if I copy paste that example in my code?

Reviewer: I don't care that much just trying to understand what was the reason to add it.

Author: There are many ways it could be handled, yes. This being the most draconian which will at least ensure that a dev and copies and pastes the code will go looking for the crash. Open to other recommendations here.
// cc @mcollina

Reviewer 2: Setting `process.exitCode` is better here, imo – or just re-throwing `err`, tbh.

Author: Ok, updated to `process.exitCode`

async function do_gzip(input, output) {
  const gzip = createGzip();
  const source = createReadStream(input);
  const destination = createWriteStream(output);
  await pipe(source, gzip, destination);
}

do_gzip('input.txt', 'input.txt.gz')
  .catch((err) => {
    console.error('An error occurred:', err);
    process.exit(1);
  });
```

It is also possible to compress or decompress data in a single step:

```js
const { deflate, unzip } = require('zlib');

const input = '.................................';
deflate(input, (err, buffer) => {
  if (err) {
    console.error('An error occurred:', err);
    process.exit(1);
  }
  console.log(buffer.toString('base64'));
});

const buffer = Buffer.from('eJzT0yMAAGTvBe8=', 'base64');
unzip(buffer, (err, buffer) => {
  if (err) {
    console.error('An error occurred:', err);
    process.exit(1);
  }
  console.log(buffer.toString());
});

// Or, Promisified

const { promisify } = require('util');
const do_unzip = promisify(unzip);

do_unzip(buffer)
  .then((buf) => console.log(buf.toString()))
  .catch((err) => {
    console.error('An error occurred:', err);
    process.exit(1);
  });
```

## Threadpool Usage

> 

Reviewer: Please do not make this softer because it was explicitly added to raise awareness against https://github.com/nodejs/node/issues/8871.

Author: The current text really doesn't say much that would be useful. If we're going to say anything then we should be much more specific and give an example of the kind of trouble users can face.

Reviewer: It says that it can cause memory fragmentation. That part was completely removed.

Author: Yes, the average user is really not going to know what that means. It's scary language without context. I've no problem with it being there but we should at least show an example of what we mean by "catastrophic memory fragmentation"

Reviewer: I'm -1 on removing it as it literally reverts https://github.com/nodejs/node/pull/20380 so it seems a step backwards to me.

Author: That's fine but we should have an example here. I'll see if I can work something up;

Author: Updated with an example

    process.exitCode = 1;
  });
```

It is also possible to compress or decompress data in a single step:

```js
const { deflate, unzip } = require('zlib');

const input = '.................................';
deflate(input, (err, buffer) => {
  if (err) {
    console.error('An error occurred:', err);
    process.exitCode = 1;
  }
  console.log(buffer.toString('base64'));
});

const buffer = Buffer.from('eJzT0yMAAGTvBe8=', 'base64');
unzip(buffer, (err, buffer) => {
  if (err) {
    console.error('An error occurred:', err);
    process.exitCode = 1;
  }
  console.log(buffer.toString());
});

// Or, Promisified

const { promisify } = require('util');
const do_unzip = promisify(unzip);

do_unzip(buffer)
  .then((buf) => console.log(buf.toString()))
  .catch((err) => {
    console.error('An error occurred:', err);
    process.exitCode = 1;
  });
```

## Threadpool Usage and Performance Considerations

All `zlib` APIs, except those that are explicitly synchronous, use the Node.js
internal threadpool. This can lead to surprising effects and performance
limitations in some applications.

Creating and using a large number of zlib objects simultaneously can cause

> significant and unrecoverable memory fragmentation errors that can

Reviewer 2: There are no errors involved here, and the memory fragmentation isn’t generally unrecoverable either – I’d just say “significant memory fragmentation“ and leave it at that, tbh. (Although that phrasing is already used below, too…)

> be catastrophic to the Node.js process.

Author: ```suggestion
Creating and using a large number of zlib objects simultaneously can cause
significant memory fragmentation.
```

```js
const zlib = require('zlib');

const payload = Buffer.from('This is some data');

// WARNING: DO NOT DO THIS!
for (let i = 0; i < 30000; ++i) {
  zlib.deflate(payload, (err, buffer) => {});
}
```

In the preceding example, 30,000 deflate instances are created concurrently. Because
of how some operating systems handle memory allocation and deallocation,
this may lead to to significant memory fragmentation.

> It is strongly recommended that the results of compression/decompression

Author: ```suggestion
It is strongly recommended that the results of compression
```

> Stability: 2 - Stable

The `zlib` module provides compression functionality implemented using Gzip,
Deflate/Inflate, and Brotli.

> It can be accessed using:

Reviewer 3: FWIW, we use this in some of our docs:

```suggestion
To access it:
```

I prefer it because it is active voice, direct, and concise. But admittedly, this is Just My Opinion.

Reviewer 3: ^^^^ Non-blocking, obviously. I also like  `Access it with:` if that's more agreeable for others.

```js
const zlib = require('zlib');
```

> Compression and decompression is built around the Node.js [Streams API][].

Reviewer 3: ```suggestion
Compression and decompression are built around the Node.js [Streams API][].
```

  const gzip = createGzip();
  const source = createReadStream(input);
  const destination = createWriteStream(output);
  await pipe(source, gzip, destination);
}

do_gzip('input.txt', 'input.txt.gz')
  .catch((err) => {
    console.error('An error occurred:', err);
    process.exitCode = 1;
  });
```

It is also possible to compress or decompress data in a single step:

```js
const { deflate, unzip } = require('zlib');

const input = '.................................';
deflate(input, (err, buffer) => {
  if (err) {
    console.error('An error occurred:', err);
    process.exitCode = 1;
  }
  console.log(buffer.toString('base64'));
});

const buffer = Buffer.from('eJzT0yMAAGTvBe8=', 'base64');
unzip(buffer, (err, buffer) => {
  if (err) {
    console.error('An error occurred:', err);
    process.exitCode = 1;
  }
  console.log(buffer.toString());
});

// Or, Promisified

const { promisify } = require('util');
const do_unzip = promisify(unzip);

do_unzip(buffer)
  .then((buf) => console.log(buf.toString()))
  .catch((err) => {
    console.error('An error occurred:', err);
    process.exitCode = 1;
  });
```

## Threadpool Usage and Performance Considerations

> All zlib APIs, except those that are explicitly synchronous, use Node.js'

Reviewer 3: ```suggestion
All `zlib` APIs, except those that are explicitly synchronous, use the Node.js
```

Reviewer 3: ^^^^^ FWIW, I actively avoid `Node.js'` and `Node.js's` because it's visually jarring and potentially confusing. But again, that's Just My Opinion and not something I'm going to block on or anything.

internal threadpool. This can lead to surprising effects and performance
limitations in some applications.

Creating and using a large number of zlib objects simultaneously can cause

> significant and unrecoverable memory fragmentation errors to occur that can

Reviewer 3: ```suggestion
significant and unrecoverable memory fragmentation errors that can
```

be catastrophic to the Node.js process.

> This problem is illustrated in the following example:

Reviewer 3: Omit?

Author: ```suggestion
```

```js
const zlib = require('zlib');

const payload = Buffer.from('This is some data');

> // Caution: do not do this

Reviewer 3: This might warrant something bolder?

```suggestion
// WARNING: DO NOT DO THIS!
```

for (let i = 0; i < 30000; ++i) {
  zlib.deflate(payload, (err, buffer) => {});
}
```

> In this example, 30,000 deflate instances are created concurrently. Because

Reviewer 3: ```suggestion
In the preceding example, 30,000 deflate instances are created concurrently. Because
```

> of how some operating systems handle memory allocation and de-allocation

Reviewer 3: ```suggestion
of how some operating systems handle memory allocation and deallocation
```

> in this case, this may lead to to significant memory fragmentation.

Reviewer 3: ```suggestion
this may lead to to significant memory fragmentation.
```

This would require the comma to be moved to the previous line after `deallocation`. I'll add that to the suggestion above, but then they need to be accepted or rejected as a pair....

It is strongly recommended that the results of compression/decompression

> operations be cached to avoid duplication of effort.

Reviewer 3: Does this just mean save the results of compression/decompression to a variable or a file rather than running the same compression/decompression twice? I'm not sure this won't confuse people more than enlighten, but I'm also not sure I have better wording. I'd be in favor of removing it, but like everything else I'm leaving in here, non-blocking, Just My Opinion, etc.

Reviewer 2: Yes – Depending on the use case, storing it in memory or in a file makes sense compared to compressing multiple times. 

(Decompression is both faster and caching the result is costlier, so I wouldn’t recommend caching for that here.)

  }
  console.log(buffer.toString());
});

// Or, Promisified

const { promisify } = require('util');
const do_unzip = promisify(unzip);

do_unzip(buffer)
  .then((buf) => console.log(buf.toString()))
  .catch((err) => {
    console.error('An error occurred:', err);
    process.exit(1);
  });
```

## Threadpool Usage

All zlib APIs, except those that are explicitly synchronous, use Node.js'
internal threadpool. This can lead to surprising effects and performance
limitations in some applications.

Specifically, creating and using a large number of zlib objects
simultaneously can cause significant and unrecoverable memory fragmentation
errors to occur that can be catastrophic to the Node.js process.

This problem is illustrated in the following example:

```js
const zlib = require('zlib');

const payload = Buffer.from('This is some data');

// Caution: do not do this
for (let i = 0; i < 30000; ++i) {
  zlib.deflate(payload, (err, buffer) => {});
}
```

In this example, 30,000 deflate instances are created concurrently and
contend with each other for execution on the limited threadpool. Because of
how the compression algorithms allocate and use memory internally, this

> contention leads to significant fragmentation of memory.

Reviewer 2: This isn’t quite right – it has almost nothing to to with how compression algorithms use or allocate memory or the fact that the threadpool is limited in size (or what that size is). It’s a platform-specific issue with the underlying system allocator.

So instead how about:

```
In this example, 30,000 deflate instances are created concurrently. Because of
how some operating systems handle memory allocation and de-allocation
in this case, this may lead to to significant memory fragmentation.
```

Author: Please take a look now :-)

> Such performance issues can be mitigated by increasing the [pool size][])

Reviewer 2: ```suggestion
Such performance issues can be mitigated by increasing the [pool size][]
```

That being said, this line is likely incorrect. As mentioned, the issue is not related to the thread pool size.

> Stability: 2 - Stable

The `zlib` module provides compression functionality implemented using Gzip,
Deflate/Inflate, and Brotli.

It can be accessed using:

```js
const zlib = require('zlib');
```

Compression and decompression is built around the Node.js [Streams API][].

Compressing or decompressing a stream (such as a file) can be accomplished by
piping the source stream through a `zlib` `Transform` stream into a destination
stream:

```js
const { createGzip } = require('zlib');
const { pipeline } = require('stream');
const {
  createReadStream,
  createWriteStream
} = require('fs');

const gzip = createGzip();
const source = createReadStream('input.txt');
const destination = createWriteStream('input.txt.gz');

pipeline(source, gzip, destination, (err) => {

>   // handle error

Reviewer 4: please handle the error, this is going to be copy-pasted a lot (I would process.exit here)

  switch (response.headers['content-encoding']) {
    case 'br':
      pipeline(response, zlib.createBrotliDecompress(), output, (err) => {
        // Handle error.
      });
      break;
    // Or, just use zlib.createUnzip() to handle both of the following cases:
    case 'gzip':
      pipeline(response, zlib.createGunzip(), output, (err) => {
        // Handle error.
      });
      break;
    case 'deflate':
      pipeline(response, zlib.createInflate(), outout, (err) => {
        // Handle error.
      });
      break;
    default:
      pipeline(response, output, (err) => {
        // Handle error.
      });

>       break;

Reviewer 4: Please handle the error in these cases. It's an http server, so we know how to handle it: return a 500.

Author: This is the client side. Will handle by printing error message and exiting the process.

  // See https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.3
  if (/\bdeflate\b/.test(acceptEncoding)) {
    response.writeHead(200, { 'Content-Encoding': 'deflate' });
    pipeline(raw, zlib.createDeflate(), response, (err) => {

>       // Handle error.

Author: @mcollina ... given that `writeHead()` has already been called and at least some of the body may have already been written, we cannot reasonably send a `500` at this point. The best we can do is terminate abruptly and log the error. Is there something else you would recommend here for the example?

Reviewer 4: Explicitly states that we do nothing for the reason you mentioned.

```js
const zlib = require('zlib');
const http = require('http');
const { pipeline } = require('stream');

http.createServer((request, response) => {
  // For the sake of simplicity, the Accept-Encoding checks are omitted.
  response.writeHead(200, { 'content-encoding': 'gzip' });
  const output = zlib.createGzip();
  pipeline(output, response, (err) => {

>     // Handle error.

Reviewer 4: We can prefill this code as well.

Author: See previous comment. What do you recommend here given that the `writeHead()` has already been called and some amount of data may have already been written to the client
